text-feature-translation/src/main/java/net/kyori/adventure/text/feature/translation/ResourceBundleTranslator.java

```diff
@@ -0,0 +1,168 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.feature.translation;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.charset.StandardCharsets;
+import java.text.MessageFormat;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+import java.util.MissingResourceException;
+import java.util.PropertyResourceBundle;
+import java.util.ResourceBundle;
+import java.util.WeakHashMap;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A resource bundle translator.
+ */
+public class ResourceBundleTranslator implements Translator {
+  private static final UTF8 UTF8 = new UTF8();
+
+  private final String resourceFormat;
+  private final Map<Locale, ResourceBundle> resources;
+  private final Locale defaultLocale;
+
+  public ResourceBundleTranslator(final @NonNull String resourceFormat, final @NonNull Locale defaultLocale) {
+    this.resourceFormat = requireNonNull(resourceFormat, "resource format");
+    this.resources = Collections.synchronizedMap(new WeakHashMap<>()); // Weak to prevent infinite locale growth
+    this.defaultLocale = requireNonNull(defaultLocale, "default locale");
+    // Eagerly register the source and system locales
+    this.registerLocale(this.defaultLocale);
+    this.registerLocale(Locale.getDefault());
+  }
+
+  /**
+   * Registers a new locale.
+   *
+   * @param locale a locale
+   * @return {@code true} if the locale has translations
+   */
+  public boolean registerLocale(final @NonNull Locale locale) {
+    if (this.resources.containsKey(locale)) return false;
+
+    final String path = MessageFormat.format(this.resourceFormat, locale.toLanguageTag());
+    final ResourceBundle resource;
+    try {
+      resource = ResourceBundle.getBundle(path, locale, UTF8);
```

Comment:
Don't ResourceBundles already have their own cache, and locale resolution? It might make more sense to override the `ResourceBundle`'s own resolution rather than introducing another layer